examples/core/rgcn/hetero_rgcn.py

```diff
@@ -0,0 +1,654 @@
+                self.hidden_size,
+                self.out_dim,
+                g.ntypes,
+                self.relation_names,
+                activation=None,
+            )
+        )
+
+    def reset_parameters(self):
+        # Reset the parameters of each layer.
+        for layer in self.layers:
+            layer.reset_parameters()
+
+    def forward(self, h, blocks):
+        for layer, block in zip(self.layers, blocks):
+            h = layer(block, h)
+        return h
+
+
+class Logger(object):
+    r"""
+    This class was taken directly from the PyG implementation and can be found
+    here: https://github.com/snap-stanford/ogb/blob/master/examples/nodeproppre
+    d/mag/logger.py
+
+    This was done to ensure that performance was measured in precisely the same
+    way
+    """
+
+    def __init__(self, runs):
+        self.results = [[] for _ in range(runs)]
+
+    def add_result(self, run, result):
+        assert len(result) == 3
+        assert run >= 0 and run < len(self.results)
+        self.results[run].append(result)
+
+    def print_statistics(self, run=None):
+        if run is not None:
+            result = 100 * th.tensor(self.results[run])
+            argmax = result[:, 1].argmax().item()
+            print(f"Run {run + 1:02d}:")
+            print(f"Highest Train: {result[:, 0].max():.2f}")
+            print(f"Highest Valid: {result[:, 1].max():.2f}")
+            print(f"  Final Train: {result[argmax, 0]:.2f}")
+            print(f"   Final Test: {result[argmax, 2]:.2f}")
+        else:
+            result = 100 * th.tensor(self.results)
+
+            best_results = []
+            for r in result:
+                train1 = r[:, 0].max().item()
+                valid = r[:, 1].max().item()
+                train2 = r[r[:, 1].argmax(), 0].item()
+                test = r[r[:, 1].argmax(), 2].item()
+                best_results.append((train1, valid, train2, test))
+
+            best_result = th.tensor(best_results)
+
+            print(f"All runs:")
+            r = best_result[:, 0]
+            print(f"Highest Train: {r.mean():.2f} ± {r.std():.2f}")
+            r = best_result[:, 1]
+            print(f"Highest Valid: {r.mean():.2f} ± {r.std():.2f}")
+            r = best_result[:, 2]
+            print(f"  Final Train: {r.mean():.2f} ± {r.std():.2f}")
+            r = best_result[:, 3]
+            print(f"   Final Test: {r.mean():.2f} ± {r.std():.2f}")
+
+
+def train(
+    g,
+    model,
+    node_embed,
+    optimizer,
+    train_loader,
+    split_idx,
+    labels,
+    logger,
+    device,
+    run,
+):
+    print("start training...")
+    category = "paper"
+
+    for epoch in range(3):
+        num_train = split_idx["train"][category].shape[0]
+        # Initialize a progress bar for visual feedback.
+        pbar = tqdm(total=num_train)
+        pbar.set_description(f"Epoch {epoch:02d}")
+        model.train()
+
+        total_loss = 0
+
+        for input_nodes, seeds, blocks in train_loader:
```

Comment:
Can we use `tqdm` to wrap the `train_loader`? Then we don't need to manage the pbar.

Reply:
This depends on whether we want to display `dataset.shape[0]`(current) or **the total number of batches**.

Reply:
I think using `tqdm` directly would be better, as `pbar` might be outside the scope of this example.